catroid/src/main/java/org/catrobat/catroid/embroidery/EmbroideryManager.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.embroidery;
+
+import android.content.res.Resources;
+import android.graphics.PointF;
+import android.util.DisplayMetrics;
+
+import org.catrobat.catroid.ProjectManager;
+import org.catrobat.catroid.common.BrickValues;
+import org.catrobat.catroid.content.XmlHeader;
+
+import java.util.ArrayList;
+
+public final class EmbroideryManager {
+	private static ArrayList<PointF> stitchPoints = new ArrayList<>();
+	private static float stitchSize = BrickValues.STITCH_SIZE * calculateScreenRatio();
+	private static int lastIndex = 0;
+
+	private EmbroideryManager() {
+
+	}
+
+	private static float calculateScreenRatio() {
+		DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
+		XmlHeader header = ProjectManager.getInstance().getCurrentProject().getXmlHeader();
+		float deviceDiagonalPixel = (float) Math.sqrt(Math.pow(metrics.widthPixels, 2) + Math.pow(metrics.heightPixels, 2));
+		float creatorDiagonalPixel = (float) Math.sqrt(Math.pow(header.getVirtualScreenWidth(), 2)
+				+ Math.pow(header.getVirtualScreenHeight(), 2));
+		return creatorDiagonalPixel / deviceDiagonalPixel;
+	}
+
+	public static void addStitchPoint(PointF point) {
```

Comment:
this isnt really code related, but stitching wise, shouldnt you be able to overstitch beginnning and ends to prevent unraveling? or is the machine dealing with this automatically?

or in another case if you would stitch something resmbling an 8, you would want to have the rising stitching path and decending stitching path both put a stitch in the center, or else you probably end up with a deliberately skipped stitch. (skipped stitch as in sewing machine lingo, which leads to problems with unraveling and easily snagged threads, aswell as could end up a problem with thread tension when done deliberately as shown here (thread tension has to be manually set and depends on stitch greatly depends on stitch lenght)).

Reply:
ah, ok, just saw below that this "isNewStitchPoint" checks only the last stitch and not if the list contains this stitch already